File: contract/contract.rs
use anchor_lang::prelude::*;
use anchor_spl::token::{self, Mint, Token, TokenAccount};
use std::result::Result;

declare_id!(""); 

#[program]
pub mod friends_fund {
    use super::*;

    pub fn initialize(
        ctx: Context<Initialize>,
        name: String,
        description: String,
        goal_amount: u64,
    ) -> Result<()> {
        ctx.accounts.fundraiser.name = name;
        ctx.accounts.fundraiser.description = description;
        ctx.accounts.fundraiser.creator = *ctx.accounts.payer.key;
        ctx.accounts.fundraiser.total_donations = 0;
        ctx.accounts.fundraiser.goal_amount = goal_amount;
        ctx.accounts.fundraiser.usdc_mint = ctx.accounts.usdc_mint.key();
        ctx.accounts.fundraiser.is_active = true;
        ctx.accounts.fundraiser.bump = *ctx.bumps.get("fundraiser").unwrap();
        Ok(())
    }

    pub fn donate(ctx: Context<Donate>, amount: u64) -> Result<()> {
        require!(ctx.accounts.fundraiser.is_active, FundraiserError::InactiveFundraiser);
        require!(amount > 0, FundraiserError::InvalidAmount);
        
        token::transfer(
            CpiContext::new(
                ctx.accounts.token_program.to_account_info(),
                token::Transfer {
                    from: ctx.accounts.donor_usdc_account.to_account_info(),
                    to: ctx.accounts.fundraiser_usdc_account.to_account_info(),
                    authority: ctx.accounts.donor.to_account_info(),
                },
            ),
            amount,
        )?;
        
        msg!("Donated {} USDC from {}", amount, ctx.accounts.donor.key());
        
        ctx.accounts.fundraiser.total_donations += amount;
        ctx.accounts.fundraiser.donors.push(*ctx.accounts.donor.key);
        
        if amount >= 100_000_000 { 
            Self::mint_reward_nft(ctx)?;
        }

        Ok(())
    }

    pub fn propose_help(ctx: Context<ProposeHelp>, proposal: String) -> Result<()> {
        require!(ctx.accounts.fundraiser.is_active, FundraiserError::InactiveFundraiser);
        
        let help_proposal = HelpProposal {
            proposer: *ctx.accounts.proposer.key,
            proposal,
            votes: 0,
        };
        
        ctx.accounts.fundraiser.help_proposals.push(help_proposal);
        Ok(())
    }

    pub fn vote_for_proposal(ctx: Context<VoteForProposal>, proposal_index: u8) -> Result<()> {
        require!(ctx.accounts.fundraiser.is_active, FundraiserError::InactiveFundraiser);
        require!(
            proposal_index < ctx.accounts.fundraiser.help_proposals.len() as u8,
            FundraiserError::InvalidProposalIndex
        );
        
        ctx.accounts.fundraiser.help_proposals[proposal_index as usize].votes += 1;
        Ok(())
    }

    pub fn close_fundraiser(ctx: Context<CloseFundraiser>) -> Result<()> {
        require!(
            ctx.accounts.creator.key() == &ctx.accounts.fundraiser.creator,
            FundraiserError::Unauthorized
        );
        
        ctx.accounts.fundraiser.is_active = false;
        Ok(())
    }

    fn mint_reward_nft(ctx: Context<Donate>) -> Result<()> {        
        msg!("Awarded NFT for Great Donat {}", ctx.accounts.donor.key());
        Ok(())
    }
}

#[error_code]
pub enum FundraiserError {
    #[msg("Fanderer is not active")]
    InactiveFundraiser,
    #[msg("Non correct Amount")]
    InvalidAmount,
    #[msg("Suppress Supply Index")]
    InvalidProposalIndex,
    #[msg("Unauthorized access")]
    Unauthorized,
}

#[derive(Accounts)]
#[instruction(name: String, description: String, goal_amount: u64)]
pub struct Initialize<'info> {
    #[account(
        init,
        payer = payer,
        space = 8 + 40 + 200 + 32 + 8 + 8 + 32 + 1 + 1 + 100 + 100,
        seeds = [b"fundraiser", name.as_bytes()],
        bump
    )]
    pub fundraiser: Account<'info, Fundraiser>,
    pub usdc_mint: Account<'info, Mint>,
    #[account(mut)]
    pub payer: Signer<'info>,
    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
pub struct Donate<'info> {
    #[account(mut)]
    pub fundraiser: Account<'info, Fundraiser>,
    #[account(
        mut,
        constraint = donor_usdc_account.mint == usdc_mint.key(),
        constraint = fundraiser_usdc_account.mint == usdc_mint.key()
    )]
    pub donor_usdc_account: Account<'info, TokenAccount>,
    #[account(mut)]
    pub fundraiser_usdc_account: Account<'info, TokenAccount>,
    #[account(signer)]
    pub donor: Signer<'info>,
    pub usdc_mint: Account<'info, Mint>,
    pub token_program: Program<'info, Token>,
    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
pub struct ProposeHelp<'info> {
    #[account(mut)]
    pub fundraiser: Account<'info, Fundraiser>,
    #[account(signer)]
    pub proposer: Signer<'info>,
}

#[derive(Accounts)]
pub struct VoteForProposal<'info> {
    #[account(mut)]
    pub fundraiser: Account<'info, Fundraiser>,
    #[account(signer)]
    pub voter: Signer<'info>,
}

#[derive(Accounts)]
pub struct CloseFundraiser<'info> {
    #[account(mut)]
    pub fundraiser: Account<'info, Fundraiser>,
    #[account(signer)]
    pub creator: Signer<'info>,
}

#[account]
pub struct Fundraiser {
    pub name: String,
    pub description: String,
    pub creator: Pubkey,
    pub total_donations: u64,
    pub goal_amount: u64,
    pub usdc_mint: Pubkey,
    pub is_active: bool,
    pub bump: u8,
    pub donors: Vec<Pubkey>,
    pub help_proposals: Vec<HelpProposal>,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone)]
pub struct HelpProposal {
    pub proposer: Pubkey,
    pub proposal: String,
    pub votes: u32,
}

File: src/lib/contract.ts
import { Program, AnchorProvider, web3 } from '@project-serum/anchor';
import { PublicKey } from '@solana/web3.js';
import IDL from './friend_fund.json';
import { TOKEN_PROGRAM_ID } from '@project-serum/anchor/dist/cjs/utils/token';

const network = process.env.NEXT_PUBLIC_SOLANA_NETWORK!;
const preflightCommitment = 'processed';
const programId = new PublicKey(process.env.NEXT_PUBLIC_PROGRAM_ID!);

const getProvider = () => {
  const connection = new web3.Connection(network, preflightCommitment);
  const wallet = (window as any).solana; 
  if (!wallet) {
    throw new Error('Wallet not connected');
  }
  return new AnchorProvider(connection, wallet, { preflightCommitment });
};

const getProgramAndProvider = () => {
  const provider = getProvider();
  const program = new Program(IDL as any, programId, provider);
  return { program, provider };
};

export const createFundraiser = async (
  name: string,
  description: string,
  goalAmount: number,
  usdcMint: PublicKey
) => {
  const { program, provider } = getProgramAndProvider();
  const [fundraiserPDA] = await PublicKey.findProgramAddress(
    [Buffer.from('fundraiser'), Buffer.from(name)],
    program.programId
  );
  
  await program.methods.initialize(name, description, goalAmount)
    .accounts({
      fundraiser: fundraiserPDA,
      usdcMint,
      payer: provider.wallet.publicKey,
      systemProgram: web3.SystemProgram.programId,
    })
    .rpc();
    
  return fundraiserPDA;
};

export const donateToFundraiser = async (
  fundraiserPDA: PublicKey,
  amount: number,
  donorUsdcAccount: PublicKey,
  fundraiserUsdcAccount: PublicKey,
  usdcMint: PublicKey
) => {
  const { program, provider } = getProgramAndProvider();
  
  await program.methods.donate(amount)
    .accounts({
      fundraiser: fundraiserPDA,
      donorUsdcAccount,
      fundraiserUsdcAccount,
      donor: provider.wallet.publicKey,
      usdcMint,
      tokenProgram: TOKEN_PROGRAM_ID,
      systemProgram: web3.SystemProgram.programId,
    })
    .rpc();
};

export const mintRewardNft = async (
  fundraiserPDA: PublicKey,
  donorUsdcAccount: PublicKey,
  fundraiserUsdcAccount: PublicKey,
  usdcMint: PublicKey
) => {
  const { program, provider } = getProgramAndProvider();
  await program.methods.mintRewardNft()
    .accounts({
      fundraiser: fundraiserPDA,
      donorUsdcAccount,
      fundraiserUsdcAccount,
      donor: provider.wallet.publicKey,
      usdcMint,
      tokenProgram: TOKEN_PROGRAM_ID,
      systemProgram: web3.SystemProgram.programId,
    })
    .rpc();
};

export const proposeHelp = async (
  fundraiserPDA: PublicKey,
  proposal: string
) => {
  const { program, provider } = getProgramAndProvider();
  await program.methods.proposeHelp(proposal)
    .accounts({
      fundraiser: fundraiserPDA,
      proposer: provider.wallet.publicKey,
    })
    .rpc();
};

export const voteForProposal = async (
  fundraiserPDA: PublicKey,
  proposalIndex: number
) => {
  const { program, provider } = getProgramAndProvider();
  await program.methods.voteForProposal(proposalIndex)
    .accounts({
      fundraiser: fundraiserPDA,
      voter: provider.wallet.publicKey,
    })
    .rpc();
};

export const closeFundraiser = async (
  fundraiserPDA: PublicKey
) => {
  const { program, provider } = getProgramAndProvider();
  await program.methods.closeFundraiser()
    .accounts({
      fundraiser: fundraiserPDA,
      creator: provider.wallet.publicKey,
    })
    .rpc();
};

